cs/test/core/ConvertTests.cs

```diff
@@ -0,0 +1,112 @@
+        public List<decimal> Numbers { get; set; }
+
+        public Decimals()
+            : this("UnitTest.Convert.Decimals", "Decimals")
+        { }
+
+        protected Decimals(string fullName, string name)
+        {
+            Price = new decimal();
+            Numbers = new List<decimal>();
+        }
+    }
+
+    public static class BondTypeAliasConverter
+    {
+        public static int ConvertToDecimalCount = 0;
+        public static int ConvertToArraySegmentCount = 0;
+
+        public static decimal Convert(ArraySegment<byte> value, decimal unused)
+        {
+            var bits = new int[value.Count / sizeof(int)];
+            Buffer.BlockCopy(value.Array, value.Offset, bits, 0, bits.Length * sizeof(int));
+
+            Interlocked.Increment(ref ConvertToDecimalCount);
+
+            return new decimal(bits);
+        }
+
+        public static ArraySegment<byte> Convert(decimal value, ArraySegment<byte> unused)
+        {
+            var bits = decimal.GetBits(value);
+            var data = new byte[bits.Length * sizeof(int)];
+            Buffer.BlockCopy(bits, 0, data, 0, data.Length);
+
+            Interlocked.Increment(ref ConvertToArraySegmentCount);
+
+            return new ArraySegment<byte>(data);
+        }
+
+        public static void Clear()
+        {
+            ConvertToDecimalCount = 0;
+            ConvertToArraySegmentCount = 0;
+        }
+    }
+
+
+    [TestFixture]
+    public class TypeAliasTests
+    {
+        [SetUp]
+        public void Setuo()
+        {
+            BondTypeAliasConverter.Clear();
+        }
+
+        [Test]
+        public void CorrectSerializeConvertCount()
+        {
+            var foo = new Decimals();
+            foo.Price = new decimal(19.91);
+
+            var outputStream = new OutputBuffer();
+            var writer = new CompactBinaryWriter<OutputBuffer>(outputStream);
+            var serializer = new Serializer<CompactBinaryWriter<OutputBuffer>>(typeof(Decimals));
+
+            serializer.Serialize(foo, writer);
+
+            Assert.AreEqual(1, BondTypeAliasConverter.ConvertToArraySegmentCount);
+            Assert.AreEqual(0, BondTypeAliasConverter.ConvertToDecimalCount);
+        }
+
+        [Test]
+        // this test currently fails
+        public void CorrectSerializeListConvertCount()
+        {
+            var foo = new Decimals();
+            foo.Price = new decimal(19.91);
+            foo.Numbers = new List<decimal> { new decimal(10.10), new decimal(11.11) };
+
+            var outputStream = new OutputBuffer();
+            var writer = new CompactBinaryWriter<OutputBuffer>(outputStream);
+            var serializer = new Serializer<CompactBinaryWriter<OutputBuffer>>(typeof(Decimals));
+
+            serializer.Serialize(foo, writer);
+
+            Assert.AreEqual(1+foo.Numbers.Count, BondTypeAliasConverter.ConvertToArraySegmentCount);
+            Assert.AreEqual(0, BondTypeAliasConverter.ConvertToDecimalCount);
+        }
```

Comment:
We could also add tests that count how many converts happen on **deserialize**, just to make sure we never regress.

Reply:
Yes, I plan to expand test coverage to cover maps, vectors, etc., and also to add deserialization tests.